### ISSUE TITLE
Regression in DSL.or(Collection) when passing empty collection

### ISSUE THREAD

**Issue by arlampin** [jOOQ/jOOQ#7584]
### Expected behavior and actual behavior:
In 3.10.x and earlier, `DSL.or(Collection<? extends Condition> conditions)` returned `DSL.falseCondition()` for an empty collection. With 3.11.0 it returns `DSL.noCondition()` (#6817), which is omitted from WHERE clause, which usually causes different query results than 3.10.x did.

### Steps to reproduce the problem (include example code if possible):

```java
List<Condition> conditions = new ArrayList<>();
Result<Record1<Integer>> one = context.selectOne().where(DSL.or(conditions)).fetch();
System.out.println(one);
```
The above returns different values for 3.10 and 3.11

### Versions:

- jOOQ: 3.11.0
- Java:
- Database (include vendor):
- OS:
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll look into whether this is an undocumented change or a regression. The behaviour is undefined in the Javadoc (which should be fixed with this issue).

I tend to think of this as a regression. The identity value for the `OR` reduction is `false`, not `noCondition()` or `true`

**Comment by arlampin** [comment]
Is this going to 3.11.1 patch?

Apart from selects, this also affects e.g. deletes: `context.deleteFrom(...).where(DSL.or(new ArrayList<>())).execute()` either deletes everything or nothing, depending on jOOQ version.

**Comment by lukaseder** [comment]
> Is this going to 3.11.1 patch?

Yes. This week, 3.11.1 is expected to be released, including a fix for this. I'll work on the fix today

> Apart from selects, this also affects e.g. deletes: context.deleteFrom(...).where(DSL.or(new ArrayList<>())).execute() either deletes everything or nothing, depending on jOOQ version.

Indeed, this isn't specific to any statement. The problem is that `DSL.or(...)` produces a `DSL.noCondition()` for empty argument collections / arrays, which it shouldn't. That is an internal optimisation that has leaked to the public API:
https://github.com/jOOQ/jOOQ/blob/version-3.11.0/jOOQ/src/main/java/org/jooq/impl/CombinedCondition.java#L83

The logic that translates empty or-connected lists of conditions to a `DSL.falseCondition()` is still present:
https://github.com/jOOQ/jOOQ/blob/version-3.11.0/jOOQ/src/main/java/org/jooq/impl/CombinedCondition.java#L147

But it is intercepted by the above optimisation

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.12. Backport for 3.11.1 scheduled: #7593. 3.11.1 will be released today. 3.10 is not affected.

Thanks again for your report!
